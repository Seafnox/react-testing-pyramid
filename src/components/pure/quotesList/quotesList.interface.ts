import { IQuote } from '../../../interfaces/IQuote';

export interface IQuotesListState {
    quotes: IQuote[];
}

export interface IQuotesProps extends IQuotesListState {
}
