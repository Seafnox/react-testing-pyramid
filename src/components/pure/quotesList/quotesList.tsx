import React, { PureComponent } from 'react';
import { IQuotesProps } from './quotesList.interface';
import { QuoteItem, QuoteText, QuoteAuthor } from './qutesList.elements';

export class QuotesList extends PureComponent<IQuotesProps> {
    render(): React.ReactElement {
        return (
            <div>
                {this.props.quotes.map(({text, author, id}) => (
                    <QuoteItem key={id} to={`/quote/${id}`}>
                        <QuoteText>{text}</QuoteText>
                        <br />
                        <QuoteAuthor>{author}</QuoteAuthor>
                    </QuoteItem>
                ))}
            </div>
        );
    }
}
