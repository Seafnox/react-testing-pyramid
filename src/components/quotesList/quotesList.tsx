import React, { Component } from 'react';
import { connect } from 'react-redux';
import { QuoteItem, QuoteText, QuoteAuthor } from './qutesList.elements';

export class QuotesListComponent extends Component<void> {
    render(): React.ReactElement {
        return (
            <div>
                <QuoteItem>
                    <QuoteText>Start</QuoteText>
                    <br />
                    <QuoteAuthor>Typing</QuoteAuthor>
                </QuoteItem>
                <QuoteItem>
                    <QuoteText>Start</QuoteText>
                    <br />
                    <QuoteAuthor>Typing</QuoteAuthor>
                </QuoteItem>
            </div>
        );
    }
}

export const QuotesList = connect()(QuotesListComponent);
