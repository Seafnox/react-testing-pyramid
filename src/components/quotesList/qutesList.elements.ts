import { Link } from 'react-router-dom';
import styled from 'styled-components';

export const QuoteItem = styled(Link)`
    margin-bottom: 10px;
    padding: 10px;
    border: 1px solid #e1e1e1;
    color: #000;
    display: block;
    text-decoration: none;
`;

export const QuoteText = styled.span`
    margin-bottom: 5px;
    font-size: 16px;
    text-decoration: underline;
`;

export const QuoteAuthor = styled.span`
    font-style: italic;
    font-size: 14px;
    color: #666;
`;