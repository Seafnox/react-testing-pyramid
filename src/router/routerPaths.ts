export const mainRoute = '/';
export const quotePageRoute = (quoteId: string): string => `/quote/${quoteId}`;
export const quoteIdParam = 'quoteId';