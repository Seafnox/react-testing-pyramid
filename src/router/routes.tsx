import React from 'react';
import { Route, Switch } from 'react-router';
import { QuotePageLoadable } from '../components/quotePage/quotePage.loadable';
import { QuotesPageLoadable } from '../components/quotesPage/quotesPage.loadable';
import { mainRoute, quotePageRoute, quoteIdParam } from './routerPaths';

export const Routes = (): JSX.Element => (
    <Switch>
        <Route exact path={mainRoute} component={QuotesPageLoadable}/>
        <Route exact path={quotePageRoute(`:${quoteIdParam}`)} component={QuotePageLoadable}/>
    </Switch>
);