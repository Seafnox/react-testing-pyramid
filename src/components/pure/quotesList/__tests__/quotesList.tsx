import React from 'react';
import { MemoryRouter } from 'react-router';
import { quotesMock } from '../../../../test-utils/mocks/qoutes.mock';
import { quotePageRoute } from '../../../../router/routerPaths';
import { QuotesList } from '../quotesList';
import { QuoteItem, QuoteText, QuoteAuthor } from '../qutesList.elements';
import { create, ReactTestRenderer, ReactTestInstance } from 'react-test-renderer';

describe('QuotesList - quotes list component', () => {
    const [firstQuote] = quotesMock;

    let container: ReactTestRenderer;
    let instance: ReactTestInstance;

    beforeEach(() => {
        container = create(<MemoryRouter><QuotesList quotes={quotesMock} /></MemoryRouter>);
        instance = container.root;

    });

    afterEach(() => {
        container.unmount();
    });

    it('Snapshot of component is match', () => {
        expect(container.toJSON()).toMatchSnapshot();
    });

    /**
     * TODO: other tests is legacy
     * @see https://www.valentinog.com/blog/testing-react/
     */
    it('Count of displayed quotes matches the input data',  () => {
        const quotesList = instance.findAllByType(QuoteItem);

        expect(quotesList.length).toBe(quotesMock.length);
    });

    it('Text of first quote is correct', () => {
        const [quoteText] = instance.findAllByType(QuoteText);
        const text = quoteText.props.children;

        expect(text).toBe(firstQuote.text);
    });

    it('Author of first quote is correct', () => {
        const [quoteAuthor] = instance.findAllByType(QuoteAuthor);
        const text = quoteAuthor.props.children;

        expect(text).toBe(firstQuote.author);
    });

    it('First quote item have link to its page', () => {
        const [quoteItem] = instance.findAllByType(QuoteItem);

        expect(quoteItem.props.to).toBe(quotePageRoute(firstQuote.id));
    });
});
