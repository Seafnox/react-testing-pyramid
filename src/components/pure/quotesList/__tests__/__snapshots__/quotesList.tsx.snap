// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuotesList - quotes list component Snapshot of component is match 1`] = `
<div
  id="quotes-list"
>
  <a
    className="sc-bdVaJa quote-item dvUrPw"
    href="/quote/1"
    onClick={[Function]}
  >
    <span
      className="sc-bwzfXH quote-text cgDlUj"
    >
      Hakuna matata!
    </span>
    <br />
    <span
      className="sc-htpNat quote-author kaHVia"
    >
      Pumba
    </span>
  </a>
  <a
    className="sc-bdVaJa quote-item dvUrPw"
    href="/quote/2"
    onClick={[Function]}
  >
    <span
      className="sc-bwzfXH quote-text cgDlUj"
    >
      Developers developers developers!
    </span>
    <br />
    <span
      className="sc-htpNat quote-author kaHVia"
    >
      Steve Ballmer
    </span>
  </a>
  <a
    className="sc-bdVaJa quote-item dvUrPw"
    href="/quote/3"
    onClick={[Function]}
  >
    <span
      className="sc-bwzfXH quote-text cgDlUj"
    >
      Just do it!
    </span>
    <br />
    <span
      className="sc-htpNat quote-author kaHVia"
    >
      Shia LaBeouf
    </span>
  </a>
</div>
`;
